server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+    if (matchingDeviceType == IOType::Input)
+        isInput = true;
+    else if (matchingDeviceType == IOType::Output)
+        isInput = false;
+
+    if (isInput) {
+        *matchingDevice = apiInfo->defaultInputDevice;
+    } else {
+        *matchingDevice = apiInfo->defaultOutputDevice;
+    }
+    if (*matchingDevice != paNoDevice)
+        fprintf(stdout, "Selecting default %s %s device\n", apiInfo->name, (isInput ? "input" : "output"));
+}
+
+// this function will select default PA devices if they are not defined
+// it will also try to check for some configuration problems
+// numIns, numOuts and sampleRate are only the requested values, can change later
+PaError SC_PortAudioDriver::CheckPaDevices(int* inDevice, int* outDevice, int numIns, int numOuts,
+                                           double sampleRate) const {
+    if (numIns && !numOuts) {
+        *outDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(
+            inDevice, sampleRate, (*inDevice != paNoDevice) ? Pa_GetDeviceInfo(*inDevice)->maxInputChannels : 0,
+            Pa_GetDefaultInputDevice(),
+            [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(&params, nullptr, sr); }, "input");
+    } else if (!numIns && numOuts) {
+        *inDevice = paNoDevice;
+        // check for requested sample rate or select the default device
+        return CheckSinglePaDevice(
+            outDevice, sampleRate, (*outDevice != paNoDevice) ? Pa_GetDeviceInfo(*outDevice)->maxOutputChannels : 0,
+            Pa_GetDefaultOutputDevice(),
+            [](PaStreamParameters& params, double sr) { return Pa_IsFormatSupported(nullptr, &params, sr); }, "output");
+    } else if (numIns && numOuts) {
+        // if one device is specified, let's try to open another one on matching api
+        SelectMatchingPaDevice(inDevice, outDevice, IOType::Input);
+        SelectMatchingPaDevice(outDevice, inDevice, IOType::Output);
+
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice
+            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi) {
+            fprintf(stdout, "Requested devices %s and %s use different API. ", GetPaDeviceName(*inDevice).c_str(),
+                    GetPaDeviceName(*outDevice).c_str());
+            *outDevice = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*inDevice)->hostApi)->defaultOutputDevice;
+            fprintf(stdout, "Setting output device to %s.\n", GetPaDeviceName(*outDevice).c_str());
+        }
+        // check for matching sampleRate or requested sample rate
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice) {
+            const auto in_parameters =
+                GetPaStreamParameters(*inDevice, Pa_GetDeviceInfo(*inDevice)->maxInputChannels, 0);
+            const auto out_parameters =
+                GetPaStreamParameters(*outDevice, Pa_GetDeviceInfo(*outDevice)->maxOutputChannels, 0);
```

Comment:
(suggestion) grr, I think these could fit on one line if you used `params` instead of `parameters`

Reply:
unfortunately `out_params` still doesn't fit in one line, so I'm gonna leave as is.

Reply:
ok, thanks for checking, i tried it in my shell and thought that would have put it under 120 chars but I guess not